File: insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/connection_service/MessageQueue.java
package info.nightscout.androidaps.plugins.pump.insight.connection_service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import info.nightscout.androidaps.plugins.pump.insight.app_layer.AppLayerMessage;

public class MessageQueue {

    MessageRequest activeRequest;
    final List<MessageRequest> messageRequests = new ArrayList<>();

    public MessageRequest getActiveRequest() {
        return activeRequest;
    }

    public void completeActiveRequest(AppLayerMessage response) {
        if (activeRequest == null) return;
        synchronized (activeRequest) {
            activeRequest.response = response;
            activeRequest.notifyAll();
        }
        activeRequest = null;
    }

    public void completeActiveRequest(Exception exception) {
        if (activeRequest == null) return;
        synchronized (activeRequest) {
            activeRequest.exception = exception;
            activeRequest.notifyAll();
        }
        activeRequest = null;
    }

    public void completePendingRequests(Exception exception) {
        for (MessageRequest messageRequest : messageRequests) {
            synchronized (messageRequest) {
                messageRequest.exception = exception;
                messageRequest.notifyAll();
            }
        }
        messageRequests.clear();
    }

    public void enqueueRequest(MessageRequest messageRequest) {
        messageRequests.add(messageRequest);
        Collections.sort(messageRequests);
    }

    public void nextRequest() {
        if (messageRequests.size() != 0) {
            activeRequest = messageRequests.get(0);
            messageRequests.remove(0);
        }
    }

    public boolean hasPendingMessages() {
        return messageRequests.size() != 0;
    }

    public void reset() {
        activeRequest = null;
        messageRequests.clear();
    }
}

File: insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/connection_service/MessageQueue.kt
package info.nightscout.androidaps.plugins.pump.insight.connection_service

import info.nightscout.androidaps.extensions.notifyAll
import info.nightscout.androidaps.plugins.pump.insight.app_layer.AppLayerMessage
import java.util.*

class MessageQueue {

    var activeRequest: MessageRequest<out AppLayerMessage>? = null
    val messageRequests: MutableList<MessageRequest<out AppLayerMessage>> = ArrayList()

    @Suppress("Unchecked_Cast")
    fun completeActiveRequest(response: AppLayerMessage) {
        if (activeRequest == null) return
        (activeRequest as MessageRequest<AppLayerMessage>?)?.let { activeRequest ->
            synchronized(activeRequest) {
                activeRequest.response = response
                activeRequest.notifyAll()
            }
        }
        activeRequest = null
    }

    fun completeActiveRequest(exception: Exception?) {
        if (activeRequest == null) return
        activeRequest?.let { activeRequest ->
            synchronized(activeRequest) {
                activeRequest.exception = exception
                activeRequest.notifyAll()
            }
        }
        activeRequest = null
    }

    fun completePendingRequests(exception: Exception?) {
        for (messageRequest in messageRequests) {
            synchronized(messageRequest) {
                messageRequest.exception = exception
                messageRequest.notifyAll()
            }
        }
        messageRequests.clear()
    }

    fun enqueueRequest(messageRequest: MessageRequest<*>) {
        messageRequests.add(messageRequest)
        messageRequests.sort()
    }

    fun nextRequest() {
        if (messageRequests.size != 0) {
            activeRequest = messageRequests[0]
            messageRequests.removeAt(0)
        }
    }

    fun hasPendingMessages(): Boolean {
        return messageRequests.size != 0
    }

    fun reset() {
        activeRequest = null
        messageRequests.clear()
    }
}